### PR TITLE
[MGS] Upload host phase2 recovery image(s)

### PR DESCRIPTION
**Pull request by jgallagher**

### CODE REVIEW

File: gateway/src/context.rs [hunk 1]
```diff
@@ -20,37 +21,20 @@ impl ServerContext {
...
-                &Arc::new(TempNoopHostPhase2RecoveryProvider),
-                log,
-            )
-            .await?,
-        );
+        // The capacity here is the maximum number of recovery images we're
+        // willing to keep in memory - we only ever expect to be serving one
+        // (whatever the most recent one is).
+        let host_phase2_provider =
+            Arc::new(InMemoryHostPhase2Provider::with_capacity(1));
```
Comment (jgallagher) [inline gateway/src/context.rs:28]: Maybe this `1` should be from the MGS config file instead? Happy to make that change if you think so.

**Review by andrewjstone — APPROVED** [review]
Ship it!

**Comment by sunshowers** [comment]
Per discussion with @jgallagher, this is the Phase 2 trampoline image (i.e. the image that contains the installinator), not the phase 2 host image.

File: gateway/src/http_entrypoints.rs [hunk 2]
```diff
@@ -913,6 +920,41 @@ async fn sp_power_state_set(
...
 }
 
+/// Upload a host phase2 image that can be served to recovering hosts via the
+/// host/SP control uart.
+///
+/// MGS caches this image in memory and is limited to a small, fixed number of
+/// images (potentially 1). Uploading a new image may evict the
+/// least-recently-requested image if our cache is already full.
+#[endpoint {
+    method = POST,
```
[the remark this answers is not in the document]
Comment (jclulow) [reply gateway/src/http_entrypoints.rs:930]: I think PUT is generally for idempotent requests made to a URL that identifies the submitted resource.  In this case it seems like there's one URL for all images that is a bit more like an RPC call, and that each submission may replace one or all existing resources, so POST seems about right to me.

[the remark this answers is not in the document]
Comment (sunshowers) [reply gateway/src/http_entrypoints.rs:930]: Oh yeah, totally missed the fact that this doesn't take an ID.

File: gateway/src/context.rs (hunk 1 above)
Comment (jgallagher) [reply gateway/src/context.rs]: Moved to config in 44443b48

**Pull request merged**